ota-lite outage (#57)
foundriesio/status: outage/2024-10-09-ota-lite.md

## OTA Lite Outage

We have a service referred to as "ota-lite" that is the user facing portion
of our device management platform.

The service was unavailable from 20:13UTC to 20:35UTC. During this time, actions
like publishing TUF Targets and managing devices (from user facing APIs)
were not possible. The device-gateway remained up during this outage.

### What Happened

On the previous day, 2024-10-08, we performed a task where we periodically
rotate secure keys and tokens used by our backend. After rotating keys, we
restarted each container affected by the key rotation to make sure they
continued to function properly. We also deleted some old service accounts
and tokens no longer used by our system.

We went to bed and slept comfortably.

> Later the next day we tried to deploy a minor change to the API, but it was failing to start up.

Reviewer: ```suggestion
The next day while trying to deploy a minor API change, the service was failing to start.
```

> We tried to rollback the change only to discover it couldn't run either.

Reviewer: ```suggestion
An attempt to rollback the change couldn't run either.
```

The failure logs pointed at the fact we were using a key that didn't exist.

The missing key was for a minor part of the system where we keep backups of

> each version of Targets.json published by our users. It's not strictly required,

Reviewer: ```suggestion
each version of Targets.json published by our customers. It's not strictly required,
```

so the quick temporary fix was to comment out this code to give us breathing room.

The root cause wound up being a combination of a bad/naive Grep expression combined with a bad variable naming.

> As a result we de facto rotated not only a secret content, but also a secret name; while the API K8s specs were still pointing to the old secret name.

Reviewer: ```suggestion
As a result, we de facto rotated not only a secret content, but also a secret name —while the API K8s specs were still pointing to the old secret name.
```

As the API was not restarted during this change, it continued to work as if nothing changed.
But, at a time of the roll out of a new API version it started failing due to the missing required secret.

### Timeline of Events

* **20:12 UTC** - We rolled out a minor change to OTA Lite.
* **20:13 UTC** - We notice the deployment has failed.

> * **20:18 UTC** - We had a first workaround. However, GCP Cloud Build had a huge cache-miss and building took minutes rather tWe have a service referred to as "ota-lite" that is the user facing portion
> of our device management platform.

Reviewer: ```suggestion
* **20:18 UTC** - We had a first workaround. However, GCP Cloud Build had a huge cache-miss and building took minutes rather than seconds.
```

The service was unavailable from 20:13UTC to 20:35UTC. During this time, actions
like publishing TUF Targets and managing devices (from user facing APIs)
were not possible. The device-gateway remained up during this outage.han seconds.
* **20:30 UTC** - We get a 2nd workaround to get the system up.
* **20:35 UTC** - The workaround was built and deployed. Things were functional again.
* **21:00 UTC** - The real fix was deployed.

## Lessons Learned

> * We need to fix GCP Cloud Build caching problem with how we build our container. In a pinch - it was our bottleneck.

Reviewer: ```suggestion

* We need to fix the GCP Cloud Build caching problem by how we build our container. In a pinch - it was our bottleneck.
```

> * We need better container health checks. The service was passing our health that just needed it to be up for about a second. However, it was crashing shortly after that.

Reviewer: ```suggestion
* We need better container health checks. The service was passing a check that just needed it to be up for about a second. However, it was crashing shortly after that.
```

> ## OTA Lite Outage

Author: api.fio refers to about 2x the functionality that just ota-lite does. I've tried to update the intro paragraph to make this nicer

> ## OTA Lite Outage

Reviewer 2: > api.fio refers to about 2x the functionality that just ota-lite does

That's fine.
You may describe what ota-lite is inside the document, just as you did.

On the title level I recommend using either `2024-10-09 Foundries.io resources outage` or `2024-10-09 api.foundries.io resources outage`.
I much prefer the first title, as the impact was much wider than just the API. For example, a part of the CI was also not working properly as well as a major part of the Web UI.

If you look at the outage history, this is what we used in similar past incidents.
Also, from the user perspective, it's easier to grasp/find this outage with a more public-facing resource name.

> ## OTA Lite Outage

Reviewer 2: You have the right for opinion; I respect that.

Could you use `2024-10-09 api.foundries.io server outage` (or smth similar) instead of `OTA Lite Outage` then?
IIUC this status is aimed at our users and not us. Let's keep our internal term `OTA Lite` for ourselves.

> OTA-lite was unavailble from 20:13UTC to 20:35UTC. During this time things

Author: force push fix

like publishing TUF Targets and managing devices (from user facing APIs)

> was not possible. The device-gateway remained up during this outage.

Author: I tried to make this nicer in the force push

### What Happened

On the previous day, 2024-10-08, we performed a task where we periodically
rotate secure keys and tokens used by our backend. After rotating keys, we
restarted each container affected by the key rotation to make sure they
continuted to function properly. We also deleted some old service accounts
and tokens no longer used by our system.

> We went to bed and slept comfortably.

Reviewer 2: ```suggestion
At this point, all our services were working just fine.
So, we went to bed and slept comfortably.
```

Not sure about the second sentence still; at least we need to reason why we _slept comfortably_.

> We went to bed and slept comfortably.

Reviewer: I think just "At this point…" is fine, and we could remove "So, we went to bed…"

> We went to bed and slept comfortably.

Author: sorry - i'd removed this sentence compeltely but it didn't make it into push after I merged suggestions. I just don't want it in the report.

> Later the next day OTA lite would not start up correctly after we tried
> to deploy a very minor change to it. We tried to rollback the change only
> to discover it couldn't run either.

Reviewer 2: I would order these events differently, to be more precise about their consequentiality:
```suggestion
Later the next day we tried to deploy a minor change to the API, but it was failing to start up.
We tried to rollback the change only to discover it couldn't run either.
```

Maybe, it is worth saying `api.foundries.io` instead of `the API` above, it's up to you.

The failure logs pointed at the fact we were using a key that didn't exist.

> The missing key was for a minor part of the system where we keep backups of
> each version of Targets.json a customer publishes. It's not stricly required,
> so the quick fix was to comment out this code to give us breathing room.

Reviewer 2: ```suggestion
The missing key was for a minor part of the system where we keep backups of
each version of Targets.json published by our users. It's not stricly required,
so the quick temporary fix was to comment out this code to give us breathing room.
```

The root cause wound up being a combination of a bad/naive Grep expression
combined with a bad variable naming. We were referencing a Kubernetes secret
in a Pod with:
```
      - name: gce-creds
        secret:
          defaultMode: 420
          secretName: ota-api-artifacts

> ```

Reviewer 2: Not sure if we should disclose our secret names to the general public... as that makes them, well, less secret.
That one sentence (the first one) is enough IMHO.

Feel free to add 1-2 sentences of some detail if your see fit; but, please, don't disclose our infrastructure internals.

Author: We can drop it but not for security reasons. If we have a threat vector where knowing the name of a secret compromises us - then we've failed.

Less text is often good. That should be the reason we drop this.

> The root cause wound up being a combination of a bad/naive Grep expression
> combined with a bad variable naming. We were referencing a Kubernetes secret
> in a Pod with:
> ```
>       - name: gce-creds
>         secret:
>           defaultMode: 420
>           secretName: ota-api-artifacts
> ```
> 
> Our grepping looked for `gce-creds` and led us to believe we needed to rotated
> the `gce-creds` key. Another Pod also happened to use this. We rotated the
> key and restarted both containers and validated things were working.
> Later, we determied the service account token the actual seecret, `ota-api-artifacts`,
> was pointing to was no longer necessary. The Pod continued to work and more
> or less fail silently. It wasn't till tried to roll out a new deployment
> did our start-up code fail due to the invalid key.

Reviewer 2: IMHO there is too much detail in this paragraph.
Can we truncate by some 80%, e.g. like below:

```suggestion
The root cause wound up being a combination of a bad/naive Grep expression combined with a bad variable naming.
As a result we de facto rotated not only a secret content, but also a secret name; while the API K8s specs were still pointing to the old secret name.
As the API was not restarted during this change, it continued to work as if nothing changed.
But, at a time of the roll out of a new API version it started failing due to the missing required secret.
```
